Support non-Future `impl Trait` return types

Currently, [the erased trait always returns `-> Pin<Box<dyn $bounds>>`](https://github.com/spastorino/dynosaur/blob/main/dynosaur_derive/src/expand.rs#L167-L202).

This does not work with traits which do not have a blanket `impl SomeTrait for Pin<Box<dyn SomeTrait>>`.

For example, returning an `impl Iterator` currently fails:

```rust
#[dynosaur::dynosaur(DynSomeTrait)]
trait SomeTrait {
    fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_;
}

struct MyImpl([u8; 4]);

impl SomeTrait for MyImpl {
    fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_ {
        return self.0.into_iter();
    }
}

fn main() {
    let mut st = DynSomeTrait::boxed(MyImpl([3,2,4,1]));
    for x in st.get_iter() {
        println!("{}", x);
    }
}
```

produces the following:

```
error[E0226]: only a single explicit lifetime bound is permitted
 --> src/main.rs:1:1
  |
1 | #[dynosaur::dynosaur(DynSomeTrait)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the attribute macro `dynosaur::dynosaur` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `Pin<Box<dyn Iterator<Item = u8>>>` is not an iterator
 --> src/main.rs:1:1
  |
1 | #[dynosaur::dynosaur(DynSomeTrait)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Pin<Box<dyn Iterator<Item = u8>>>` is not an iterator
  |
  = help: the trait `Iterator` is not implemented for `Pin<Box<dyn Iterator<Item = u8>>>`

Some errors have detailed explanations: E0226, E0277.
For more information about an error, try `rustc --explain E0226`.
error: could not compile `testdyno` (bin "testdyno") due to 2 previous errors
```

Note: there's also an unrelated lifetime error due to the `+ '_` bound. The key bit is "`Pin<Box<dyn Iterator<Item = u8>>` is not an iterator" piece.

--- a/dynosaur/tests/pass/basic-rpitit.stdout
+++ b/dynosaur/tests/pass/basic-rpitit.stdout
@@ -13,16 +13,15 @@ mod _dynosaur_macro_dynmytrait {
     use super::*;
     trait ErasedMyTrait {
         fn foo<'life0, 'dynosaur>(&'life0 self)
-        -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>>
+        -> Box<dyn Send + 'dynosaur>
         where
         'life0: 'dynosaur,
         Self: 'dynosaur;
     }
     impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
-        fn foo<'life0, 'dynosaur>(&'life0 self)
-            -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>> where
-            'life0: 'dynosaur, Self: 'dynosaur {
-            Box::pin(<Self as MyTrait>::foo(self))
+        fn foo<'life0, 'dynosaur>(&'life0 self) -> Box<dyn Send + 'dynosaur>
+            where 'life0: 'dynosaur, Self: 'dynosaur {
+            Box::new(<Self as MyTrait>::foo(self))
         }
     }
     #[repr(transparent)]
@@ -31,10 +30,10 @@ mod _dynosaur_macro_dynmytrait {
     }
     impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
         fn foo(&self) -> impl Send {
-            let fut: ::core::pin::Pin<Box<dyn Send + '_>> = self.ptr.foo();
-            let fut: ::core::pin::Pin<Box<dyn Send + 'static>> =
-                unsafe { ::core::mem::transmute(fut) };
-            fut
+            let ret: Box<dyn Send + '_> = self.ptr.foo();
+            let ret: Box<dyn Send + '_> =
+                unsafe { ::core::mem::transmute(ret) };
+            ret
         }
     }
     impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {

--- a/dynosaur/tests/pass/default-method-trait.stdout
+++ b/dynosaur/tests/pass/default-method-trait.stdout
@@ -12,16 +12,16 @@ mod _dynosaur_macro_dynmytrait {
     use super::*;
     trait ErasedMyTrait {
         fn foo<'life0, 'dynosaur>(&'life0 mut self)
-        -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>>
+        -> Box<dyn Send + 'dynosaur>
         where
         'life0: 'dynosaur,
         Self: 'dynosaur;
     }
     impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
         fn foo<'life0, 'dynosaur>(&'life0 mut self)
-            -> ::core::pin::Pin<Box<dyn Send + 'dynosaur>> where
-            'life0: 'dynosaur, Self: 'dynosaur {
-            Box::pin(<Self as MyTrait>::foo(self))
+            -> Box<dyn Send + 'dynosaur> where 'life0: 'dynosaur,
+            Self: 'dynosaur {
+            Box::new(<Self as MyTrait>::foo(self))
         }
     }
     #[repr(transparent)]
@@ -30,10 +30,10 @@ mod _dynosaur_macro_dynmytrait {
     }
     impl<'dynosaur_struct> MyTrait for DynMyTrait<'dynosaur_struct> {
         fn foo(&mut self) -> impl Send {
-            let fut: ::core::pin::Pin<Box<dyn Send + '_>> = self.ptr.foo();
-            let fut: ::core::pin::Pin<Box<dyn Send + 'static>> =
-                unsafe { ::core::mem::transmute(fut) };
-            fut
+            let ret: Box<dyn Send + '_> = self.ptr.foo();
+            let ret: Box<dyn Send + '_> =
+                unsafe { ::core::mem::transmute(ret) };
+            ret
         }
     }
     impl<'dynosaur_struct> DynMyTrait<'dynosaur_struct> {

--- a/dynosaur/tests/pass/non-future-impl-traits.rs
+++ b/dynosaur/tests/pass/non-future-impl-traits.rs
@@ -1,0 +1,19 @@
+#[dynosaur::dynosaur(DynSomeTrait)]
+trait SomeTrait {
+    fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_;
+}
+
+struct MyImpl([u8; 4]);
+
+impl SomeTrait for MyImpl {
+    fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_ {
+        return self.0.into_iter();
+    }
+}
+
+fn main() {
+    let mut st = DynSomeTrait::boxed(MyImpl([3,2,4,1]));
+    for x in st.get_iter() {
+        println!("{}", x);
+    }
+}

--- a/dynosaur/tests/pass/non-future-impl-traits.stdout
+++ b/dynosaur/tests/pass/non-future-impl-traits.stdout
@@ -1,0 +1,73 @@
+#![feature(prelude_import)]
+#[prelude_import]
+use std::prelude::rust_2021::*;
+#[macro_use]
+extern crate std;
+trait SomeTrait {
+    fn get_iter(&mut self)
+    -> impl Iterator<Item = u8> + '_;
+}
+mod _dynosaur_macro_dynsometrait {
+    use super::*;
+    trait ErasedSomeTrait {
+        fn get_iter<'life0, 'dynosaur>(&'life0 mut self)
+        -> Box<dyn Iterator<Item = u8> + 'dynosaur>
+        where
+        'life0: 'dynosaur,
+        Self: 'dynosaur;
+    }
+    impl<DYNOSAUR: SomeTrait> ErasedSomeTrait for DYNOSAUR {
+        fn get_iter<'life0, 'dynosaur>(&'life0 mut self)
+            -> Box<dyn Iterator<Item = u8> + 'dynosaur> where
+            'life0: 'dynosaur, Self: 'dynosaur {
+            Box::new(<Self as SomeTrait>::get_iter(self))
+        }
+    }
+    #[repr(transparent)]
+    pub struct DynSomeTrait<'dynosaur_struct> {
+        ptr: dyn ErasedSomeTrait + 'dynosaur_struct,
+    }
+    impl<'dynosaur_struct> SomeTrait for DynSomeTrait<'dynosaur_struct> {
+        fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_ {
+            let ret: Box<dyn Iterator<Item = u8> + '_> = self.ptr.get_iter();
+            let ret: Box<dyn Iterator<Item = u8> + '_> =
+                unsafe { ::core::mem::transmute(ret) };
+            ret
+        }
+    }
+    impl<'dynosaur_struct> DynSomeTrait<'dynosaur_struct> {
+        pub fn boxed(value: impl SomeTrait + 'dynosaur_struct)
+            -> Box<DynSomeTrait<'dynosaur_struct>> {
+            let value = Box::new(value);
+            let value: Box<dyn ErasedSomeTrait + 'dynosaur_struct> = value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn from_ref(value: &(impl SomeTrait + 'dynosaur_struct))
+            -> &DynSomeTrait<'dynosaur_struct> {
+            let value: &(dyn ErasedSomeTrait + 'dynosaur_struct) = &*value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+        pub fn from_mut(value: &mut (impl SomeTrait + 'dynosaur_struct))
+            -> &mut DynSomeTrait<'dynosaur_struct> {
+            let value: &mut (dyn ErasedSomeTrait + 'dynosaur_struct) =
+                &mut *value;
+            unsafe { ::core::mem::transmute(value) }
+        }
+    }
+}
+use _dynosaur_macro_dynsometrait::DynSomeTrait;
+
+struct MyImpl([u8; 4]);
+
+impl SomeTrait for MyImpl {
+    fn get_iter(&mut self) -> impl Iterator<Item = u8> + '_ {
+        return self.0.into_iter();
+    }
+}
+
+fn main() {
+    let mut st = DynSomeTrait::boxed(MyImpl([3, 2, 4, 1]));
+    for x in st.get_iter() {
+        { ::std::io::_print(format_args!("{0}\n", x)); };
+    }
+}

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -1,6 +1,6 @@
 use crate::lifetime::{used_lifetimes, AddLifetimeToImplTrait, CollectLifetimes};
 use crate::receiver::has_self_in_sig;
-use crate::where_clauses::where_clause_or_default;
+use crate::where_clauses::{has_where_self_sized, where_clause_or_default};
 use proc_macro2::{Span, TokenStream};
 use quote::quote;
 use std::mem;
@@ -8,8 +8,8 @@ use syn::punctuated::Punctuated;
 use syn::token::RArrow;
 use syn::visit_mut::VisitMut;
 use syn::{
-    parse_quote, parse_quote_spanned, Error, FnArg, GenericParam, Generics, Pat, PatType,
-    ReturnType, Signature, Token, TraitItemFn, Type, TypeImplTrait,
+    parse_quote, parse_quote_spanned, Error, FnArg, GenericParam, Generics, ItemTrait, Pat,
+    PatType, ReturnType, Signature, Token, TraitItemFn, Type, TypeImplTrait, TypeParamBound,
 };
 
 /// Expands the signature of each function on the trait, converting async fn into fn with return
@@ -39,16 +39,53 @@ use syn::{
 pub(crate) fn expand_fn_sig(item_trait_generics: &Generics, trait_item_fn: &mut TraitItemFn) {
     let sig = &mut trait_item_fn.sig;
 
-    if is_async_or_rpit(sig) {
+    if is_async(sig) {
         expand_fn_input(item_trait_generics, sig);
         expand_sig_ret_ty_to_pin_box(sig);
+    } else if is_rpit(sig) {
+        expand_fn_input(item_trait_generics, sig);
+        expand_sig_ret_ty_to_box(sig);
     }
 
     // Remove default method if any for the erased trait
     trait_item_fn.default = None;
 }
 
-pub(crate) fn is_async_or_rpit(sig: &Signature) -> bool {
+pub(crate) fn is_async(sig: &Signature) -> bool {
+    match sig {
+        Signature {
+            asyncness: Some(_), ..
+        } => true,
+        Signature {
+            asyncness: None,
+            output: ReturnType::Type(_, ret),
+            ..
+        } => {
+            if let Type::ImplTrait(type_impl_trait) = &**ret {
+                type_impl_trait
+                    .bounds
+                    .iter()
+                    .find(|bound| match bound {
+                        TypeParamBound::Trait(trait_bound) => {
+                            let segments = &trait_bound.path.segments;
+
+                            segments.len() == 3
+                                && (segments[0].ident == "core" || segments[0].ident == "std")
+                                && segments[1].ident == "future"
+                                && segments[2].ident == "Future"
+                        }
+                        _ => false,
+                    })
+                    .is_some()
+            } else {
+                false
+            }
+        }
+        _ => false,
+    }
+}
+
+pub(crate) fn is_rpit(sig: &Signature) -> bool {
     match sig {
         Signature {
             asyncness: Some(_), ..
@@ -148,6 +185,14 @@ pub(crate) fn expand_sig_ret_ty_to_pin_box(sig: &mut Signature) {
     sig.output = parse_quote! { #arrow ::core::pin::Pin<Box<dyn #ret + 'dynosaur>> };
 }
 
+pub(crate) fn expand_sig_ret_ty_to_box(sig: &mut Signature) {
+    let (arrow, ret) = expand_arrow_ret_ty(sig);
+    if let Some(asyncness) = sig.asyncness.take() {
+        sig.fn_token.span = asyncness.span;
+    }
+    sig.output = parse_quote! { #arrow Box<dyn #ret + 'dynosaur> };
+}
+
 pub(crate) fn expand_sig_ret_ty_to_rpit(sig: &mut Signature) {
     let (arrow, ret) = expand_arrow_ret_ty(sig);
     if let Some(asyncness) = sig.asyncness.take() {
@@ -188,6 +233,84 @@ pub(crate) fn expand_invoke_args(sig: &Signature, ufc: bool) -> Vec<TokenStream>
     args
 }
 
+pub(crate) fn expand_blanket_impl_fn(
+    item_trait: &ItemTrait,
+    trait_item_fn: &mut TraitItemFn,
+) -> TokenStream {
+    let is_async = is_async(&trait_item_fn.sig);
+    let is_rpit = is_rpit(&trait_item_fn.sig);
+
+    expand_fn_sig(&item_trait.generics, trait_item_fn);
+    let sig = &trait_item_fn.sig;
+
+    let trait_ident = &item_trait.ident;
+    let (_, trait_generics, _) = &item_trait.generics.split_for_impl();
+    let ident = &sig.ident;
+    let args = expand_invoke_args(sig, false);
+    let value = quote! { <Self as #trait_ident #trait_generics>::#ident(#(#args),*) };
+
+    let value = if is_async {
+        quote! {
+            Box::pin(#value)
+        }
+    } else if is_rpit {
+        quote! {
+            Box::new(#value)
+        }
+    } else {
+        value
+    };
+
+    quote! {
+        #sig {
+            #value
+        }
+    }
+}
+
+pub(crate) fn expand_dyn_struct_fn(sig: &Signature) -> TokenStream {
+    if has_where_self_sized(&sig) {
+        quote! {
+            #sig {
+                unreachable!()
+            }
+        }
+    } else {
+        let ident = &sig.ident;
+        let args = expand_invoke_args(&sig, true);
+
+        if is_async(&sig) {
+            let ret = expand_ret_ty(&sig);
+            let mut sig = sig.clone();
+            expand_sig_ret_ty_to_rpit(&mut sig);
+
+            quote! {
+                #sig {
+                    let fut: ::core::pin::Pin<Box<dyn #ret + '_>> = self.ptr.#ident(#(#args),*);
+                    let fut: ::core::pin::Pin<Box<dyn #ret + 'static>> = unsafe { ::core::mem::transmute(fut) };
+                    fut
+                }
+            }
+        } else if is_rpit(&sig) {
+            let ret = expand_ret_ty(&sig);
+
+            quote! {
+                #sig {
+                    let ret: Box<dyn #ret + '_> = self.ptr.#ident(#(#args),*);
+                    let ret: Box<dyn #ret + '_> = unsafe { ::core::mem::transmute(ret) };
+                    ret
+                }
+            }
+        } else {
+            quote! {
+                #sig {
+                    self.ptr.#ident(#(#args),*)
+                }
+            }
+        }
+    }
+}
+
 fn expand_arrow_ret_ty(sig: &Signature) -> (RArrow, TokenStream) {
     match (sig.asyncness.is_some(), &sig.output) {
         (true, ReturnType::Default) => {
@@ -201,7 +324,15 @@ fn expand_arrow_ret_ty(sig: &Signature) -> (RArrow, TokenStream) {
         }
         (false, ReturnType::Type(arrow, ret)) => {
             if let Type::ImplTrait(TypeImplTrait { bounds, .. }) = &**ret {
-                return (*arrow, quote!(#bounds));
+                let mut ret_bounds: Punctuated<&TypeParamBound, Token![+]> = Punctuated::new();
+
+                for bound in bounds {
+                    if !matches!(bound, TypeParamBound::Lifetime(_)) {
+                        ret_bounds.push(bound);
+                    }
+                }
+
+                return (*arrow, quote! { #ret_bounds });
             }
         }
         _ => {}

--- a/dynosaur_derive/src/expand.rs
+++ b/dynosaur_derive/src/expand.rs
@@ -1,5 +1,6 @@
 use crate::lifetime::{used_lifetimes, AddLifetimeToImplTrait, CollectLifetimes};
 use crate::receiver::has_self_in_sig;
+use crate::sig::{is_async, is_rpit};
 use crate::where_clauses::{has_where_self_sized, where_clause_or_default};
 use proc_macro2::{Span, TokenStream};
 use quote::quote;
@@ -49,56 +50,6 @@ pub(crate) fn expand_fn_sig(item_trait_generics: &Generics, trait_item_fn: &mut 
 
     // Remove default method if any for the erased trait
     trait_item_fn.default = None;
-}
-
-pub(crate) fn is_async(sig: &Signature) -> bool {
-    match sig {
-        Signature {
-            asyncness: Some(_), ..
-        } => true,
-        Signature {
-            asyncness: None,
-            output: ReturnType::Type(_, ret),
-            ..
-        } => {
-            if let Type::ImplTrait(type_impl_trait) = &**ret {
-                type_impl_trait
-                    .bounds
-                    .iter()
-                    .find(|bound| match bound {
-                        TypeParamBound::Trait(trait_bound) => {
-                            let segments = &trait_bound.path.segments;
-
-                            segments.len() == 3
-                                && (segments[0].ident == "core" || segments[0].ident == "std")
-                                && segments[1].ident == "future"
-                                && segments[2].ident == "Future"
-                        }
-                        _ => false,
-                    })
-                    .is_some()
-            } else {
-                false
-            }
-        }
-        _ => false,
-    }
-}
-
-pub(crate) fn is_rpit(sig: &Signature) -> bool {
-    match sig {
-        Signature {
-            asyncness: Some(_), ..
-        } => true,
-        Signature {
-            asyncness: None,
-            output: ReturnType::Type(_, ret),
-            ..
-        } => {
-            matches!(**ret, Type::ImplTrait(_))
-        }
-        _ => false,
-    }
 }
 
 fn expand_fn_input(item_trait_generics: &Generics, sig: &mut Signature) {

--- a/dynosaur_derive/src/lib.rs
+++ b/dynosaur_derive/src/lib.rs
@@ -1,6 +1,4 @@
-use expand::{
-    expand_fn_sig, expand_invoke_args, expand_ret_ty, expand_sig_ret_ty_to_rpit, is_async_or_rpit,
-};
+use expand::{expand_blanket_impl_fn, expand_dyn_struct_fn, expand_fn_sig};
 use proc_macro2::{Span, TokenStream};
 use quote::quote;
 use syn::{
@@ -13,7 +11,6 @@ use syn::{
 use traits::{
     dyn_compatible_items, struct_trait_params, trait_item_erased_name, StructTraitParams,
 };
-use where_clauses::has_where_self_sized;
 
 mod expand;
 mod lifetime;
@@ -229,27 +226,7 @@ fn mk_erased_trait_blanket_impl(item_trait: &ItemTrait) -> TokenStream {
                     }
                 }
                 TraitItem::Fn(mut trait_item_fn) => {
-                    let is_async_or_rpit = is_async_or_rpit(&trait_item_fn.sig);
-
-                    expand_fn_sig(&item_trait.generics, &mut trait_item_fn);
-
-                    let sig = &trait_item_fn.sig;
-                    let ident = &sig.ident;
-                    let args = expand_invoke_args(sig, false);
-
-                    if is_async_or_rpit {
-                        quote! {
-                            #sig {
-                                Box::pin(<Self as #trait_ident #trait_generics>::#ident(#(#args),*))
-                            }
-                        }
-                    } else {
-                        quote! {
-                            #sig {
-                                <Self as #trait_ident #trait_generics>::#ident(#(#args),*)
-                            }
-                        }
-                    }
+                    expand_blanket_impl_fn(item_trait, &mut trait_item_fn)
                 }
                 TraitItem::Type(TraitItemType { ident, generics, .. }) => {
                     let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
@@ -298,58 +275,27 @@ fn mk_dyn_struct_impl_item(struct_ident: &Ident, item_trait: &ItemTrait) -> Toke
     let item_trait_ident = &item_trait.ident;
     let (_, trait_generics, where_clause) = &item_trait.generics.split_for_impl();
 
-    let items = item_trait.items.iter().map(|item| {
-        match item {
-            TraitItem::Const(TraitItemConst {
-                ident,
-                generics,
-                ty,
-                ..
-            }) => {
-                quote! {
-                    const #ident #generics: #ty = <Self as #item_trait_ident #trait_generics>::#ident;
-                }
+    let items = item_trait.items.iter().map(|item| match item {
+        TraitItem::Const(TraitItemConst {
+            ident,
+            generics,
+            ty,
+            ..
+        }) => {
+            quote! {
+                const #ident #generics: #ty = <Self as #item_trait_ident #trait_generics>::#ident;
             }
-            TraitItem::Fn(TraitItemFn { sig, .. }) => {
-                if has_where_self_sized(&sig) {
-                    quote! {
-                        #sig {
-                            unreachable!()
-                        }
-                    }
-                } else {
-                    let ident = &sig.ident;
-                    let args = expand_invoke_args(&sig, true);
-
-                    if is_async_or_rpit(&sig) {
-                        let ret = expand_ret_ty(&sig);
-                        let mut sig = sig.clone();
-                        expand_sig_ret_ty_to_rpit(&mut sig);
-
-                        quote! {
-                            #sig {
-                                let fut: ::core::pin::Pin<Box<dyn #ret + '_>> = self.ptr.#ident(#(#args),*);
-                                let fut: ::core::pin::Pin<Box<dyn #ret + 'static>> = unsafe { ::core::mem::transmute(fut) };
-                                fut
-                            }
-                        }
-                    } else {
-                        quote! {
-                            #sig {
-                                self.ptr.#ident(#(#args),*)
-                            }
-                        }
-                    }
-                }
-            }
-            TraitItem::Type(TraitItemType { ident, generics, .. }) => {
-                let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
-                quote! {
-                    type #ident #impl_generics = #ident #ty_generics #where_clause;
-                }
-            }
-            _ => Error::new_spanned(item, "unsupported item type").into_compile_error(),
         }
+        TraitItem::Fn(TraitItemFn { sig, .. }) => expand_dyn_struct_fn(sig),
+        TraitItem::Type(TraitItemType {
+            ident, generics, ..
+        }) => {
+            let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+            quote! {
+                type #ident #impl_generics = #ident #ty_generics #where_clause;
+            }
+        }
+        _ => Error::new_spanned(item, "unsupported item type").into_compile_error(),
     });
 
     let StructTraitParams {

--- a/dynosaur_derive/src/lib.rs
+++ b/dynosaur_derive/src/lib.rs
@@ -15,6 +15,7 @@ use traits::{
 mod expand;
 mod lifetime;
 mod receiver;
+mod sig;
 mod traits;
 mod where_clauses;
 

--- a/dynosaur_derive/src/sig.rs
+++ b/dynosaur_derive/src/sig.rs
@@ -1,0 +1,51 @@
+use syn::{ReturnType, Signature, Type, TypeParamBound};
+
+pub(crate) fn is_async(sig: &Signature) -> bool {
+    match sig {
+        Signature {
+            asyncness: Some(_), ..
+        } => true,
+        Signature {
+            asyncness: None,
+            output: ReturnType::Type(_, ret),
+            ..
+        } => {
+            if let Type::ImplTrait(type_impl_trait) = &**ret {
+                type_impl_trait
+                    .bounds
+                    .iter()
+                    .find(|bound| match bound {
+                        TypeParamBound::Trait(trait_bound) => {
+                            let segments = &trait_bound.path.segments;
+
+                            segments.len() == 3
+                                && (segments[0].ident == "core" || segments[0].ident == "std")
+                                && segments[1].ident == "future"
+                                && segments[2].ident == "Future"
+                        }
+                        _ => false,
+                    })
+                    .is_some()
+            } else {
+                false
+            }
+        }
+        _ => false,
+    }
+}
+
+pub(crate) fn is_rpit(sig: &Signature) -> bool {
+    match sig {
+        Signature {
+            asyncness: Some(_), ..
+        } => true,
+        Signature {
+            asyncness: None,
+            output: ReturnType::Type(_, ret),
+            ..
+        } => {
+            matches!(**ret, Type::ImplTrait(_))
+        }
+        _ => false,
+    }
+}

